Per broker annotation and label template (#218)
strimzi/proposals: 140-per-broker-annotation-and-label-templates.md

              external-dns.alpha.kubernetes.io/hostname: kafka-bootstrap.example.com.
          brokers:
            - broker: 0
              annotations:
                external-dns.alpha.kubernetes.io/hostname: kafka-0.example.com.
              labels:
                dns.zone: primary
            - broker: 1
              annotations:
                external-dns.alpha.kubernetes.io/hostname: kafka-1.example.com.
              labels:
                dns.zone: primary
            - broker: 2
              annotations:
                external-dns.alpha.kubernetes.io/hostname: kafka-2.example.com.
              labels:
                dns.zone: primary
```

This works, but it becomes repetitive and error-prone as clusters grow or when node IDs are not simple `0,1,2` sequences.
That is especially visible in KRaft clusters using node pools, where users often need to think in terms of node IDs and pod names rather than replica ordinals.

## Motivation

In most real deployments, per-broker annotations and labels follow a predictable pattern.
Examples include:

* `external-dns.alpha.kubernetes.io/hostname: kafka-{nodeId}.example.com.`
* `my-company.io/pod-name: {nodePodName}`

> * `topology.strimzi.io/broker-id: {nodeId}`

Reviewer: Here as well as elsewhere -> any `strimzi.io` labels and annotations should be set by Strimzi it self. Not by the user. So I would not use them as the example. Just use some custom ones as the one above for example, or some real live examples such as the External DNS one.

Requiring users to repeat these values broker by broker has several drawbacks:

* the `Kafka` custom resource becomes unnecessarily long
* copy-paste mistakes are easy to make
* scaling up requires editing the listener configuration first
* gaps in broker IDs make manual maintenance harder
* users cannot express “same pattern for all brokers, override only one broker if needed” cleanly

Strimzi already solved this problem for hostnames through `hostTemplate` and `advertisedHostTemplate`.
The same approach fits per-broker annotations and labels well.

## Proposal

This proposal suggests adding two new listener-level fields under `spec.kafka.listeners[].configuration`:

* `annotationsTemplate`

> * `labelsTemplate`

Reviewer: Should this be `perBrokerAnnotations` and `perBrokerLabels`? To make it clear it does not appyl to bootstrap resources. The existing templates are not prefixed that wy, but there it is a bit more obvious hows they are used.

Author: Agreed. How about using `perBrokerAnnotationsTemplate` and `perBrokerLabelsTemplate`. That way its consistent with the other template naming conventions e.g. `advertisedHostTemplate` while preserving the explicit nature of the variable name

Reviewer: That sounds good. Not sure why I left the template out.

Both fields will be maps of string keys to string values.
They define default per-broker metadata that Strimzi renders for each broker resource.

### Template variables

The template values support the same variables already used by listener host templating:

* `{nodeId}` → Kafka node ID
* `{nodePodName}` → Kubernetes pod name for the Kafka node

Only values are templated.
Keys are copied as-is.

For example:

```yaml
spec:
  kafka:
    listeners:
      - name: external
        port: 9094
        type: loadbalancer
        tls: true
        configuration:
          annotationsTemplate:
            external-dns.alpha.kubernetes.io/hostname: kafka-{nodeId}.example.com.
            my-company.io/pod-name: "{nodePodName}"
          labelsTemplate:
            strimzi.io/node-id: "{nodeId}"

>             strimzi.io/pod-name: "{nodePodName}"

Reviewer: As mentioned above -> we should not mix `strimzi.io` labels here.

```

For a node with ID `12` and pod name `my-cluster-brokers-12`, Strimzi would render:

```yaml
annotations:
  external-dns.alpha.kubernetes.io/hostname: kafka-12.example.com.
  my-company.io/pod-name: my-cluster-brokers-12
labels:
  strimzi.io/node-id: "12"
  strimzi.io/pod-name: my-cluster-brokers-12
```

### Where the templates apply

The new template fields apply only to per-broker resources.
They do not affect bootstrap resources.

They should be used everywhere Strimzi currently consumes per-broker annotations and labels from listener configuration, namely:

* per-broker `Service` resources
* per-broker `Route` resources
* per-broker `Ingress` resources

Bootstrap metadata remains configured explicitly through `configuration.bootstrap.annotations` and `configuration.bootstrap.labels`.

### Conflict handling and precedence

Users should be able to set a default pattern and still override specific brokers.
So the precedence should be:

1. render `annotationsTemplate` / `labelsTemplate`
2. overlay `configuration.brokers[].annotations` / `configuration.brokers[].labels`

> 3. use the merged result on the generated resource

Reviewer: How would users _unset_ a label or annotation in this model? I think it would be better to instead of merging the labels and annotations to do it the way that:
* If `configuration.brokers[].annotations` / `configuration.brokers[].labels` are set, they will be used
* When they are unset, `annotationsTemplate` / `labelsTemplate` will be used

I think it makes more sense that way and gives more flexibility (users who already set `configuration.brokers[].annotations` / `configuration.brokers[].labels` for given broker can easily set there everything)

Author: Good suggestion. It would indeed give more flexibility.

Author: From my perspective, I was thinking about the merge approach because its great for _defaults + small broker-specific additions_ but weak for the clarity of hidden changes that can happen use case. 

I don't have strong opinions on this one but i do agree the unset behavior is cleaner when choosing only between the template or specific additions.

This means broker-specific values always win over template-generated values.

For example:

```yaml
spec:
  kafka:
    listeners:
      - name: external
        port: 9094
        type: loadbalancer
        tls: true
        configuration:
          annotationsTemplate:
            external-dns.alpha.kubernetes.io/hostname: kafka-{nodeId}.example.com.
            external-dns.alpha.kubernetes.io/ttl: "60"
          labelsTemplate:
            dns.zone: primary
            strimzi.io/node-id: "{nodeId}"
          brokers:
            - broker: 1
              annotations:
                external-dns.alpha.kubernetes.io/hostname: special-broker.example.com.
              labels:
                dns.zone: secondary
```

The resulting metadata would be:

* broker `0`

```yaml
annotations:
  external-dns.alpha.kubernetes.io/hostname: kafka-0.example.com.
  external-dns.alpha.kubernetes.io/ttl: "60"
labels:
  dns.zone: primary
  strimzi.io/node-id: "0"
```

* broker `1`

```yaml
annotations:
  external-dns.alpha.kubernetes.io/hostname: special-broker.example.com.
  external-dns.alpha.kubernetes.io/ttl: "60"
labels:
  dns.zone: secondary
  strimzi.io/node-id: "1"
```

> This behavior is more flexible than using the template only when the per-broker map is completely absent.

Reviewer: As explained above, it is not because you cannot unset values.

#### Listener utility changes

`ListenersUtils` currently resolves per-broker metadata through helper methods such as `brokerAnnotations(listener, nodeId)` and `brokerLabels(listener, nodeId)`.
These helpers should be extended to:

* render the template maps for the current `NodeRef`
* merge rendered template values with explicit broker values
* keep explicit broker values as the winner on key conflicts

A practical implementation would:

* introduce a helper to render a templated string value using `{nodeId}` and `{nodePodName}`
* introduce a helper to render all values in a `Map<String, String>`
* add `brokerAnnotations(listener, node)` and `brokerLabels(listener, node)` overloads
* migrate the existing Kafka resource generation code to use the `NodeRef` overloads

#### Kafka cluster generation changes

The per-broker resource generation in `KafkaCluster` already routes all per-broker listener metadata through `ListenersUtils`.
Therefore the feature can be implemented centrally without duplicating logic in each resource builder.

The affected generated resources are:

* per-broker external services
* per-broker routes

> * per-broker ingresses

Reviewer: The TLSRoutes resources would be affected as well (not implemented yet, but already approved as a proposal). I guess it should also apply to `cluster-ip` type per-broker services in `type: cluster-ip` listeners?

Author: `cluster-ip` yes.
Do you prefer i leave a small note that it will support TLSRoutes when implemented?

              labels:
                exposure-type: premium
```

This keeps the common case short while preserving precise control where needed.

## Affected/not affected projects

Affected:

* `strimzi-kafka-operator` API model
* `strimzi-kafka-operator` Cluster Operator listener validation and resource generation
* generated CRD documentation for listener configuration
* user-facing listener documentation

Not affected:

* Topic Operator behavior
* User Operator behavior
* Kafka protocol behavior itself
* bootstrap listener metadata behavior

## Compatibility

This proposal is fully backwards compatible.

* Existing configurations using only `brokers[].annotations` and `brokers[].labels` continue to work unchanged.
* Existing bootstrap metadata configuration is unchanged.
* The change is additive and introduces no breaking API behavior.
* Users can adopt the templates gradually.

## Rejected alternatives

### Apply templates only when `brokers[].annotations` or `brokers[].labels` are absent

This was rejected because it makes partial overrides awkward.
Users often want a common default plus a targeted override for one broker.

> Merging template values first and then overlaying explicit broker metadata is more useful.

Reviewer: Sorry, I would argue that the partial overrides are awkward 😉

          brokers:
            - broker: 0
              annotations:
                external-dns.alpha.kubernetes.io/hostname: kafka-0.example.com.
              labels:
                dns.zone: primary
            - broker: 1
              annotations:
                external-dns.alpha.kubernetes.io/hostname: kafka-1.example.com.
              labels:
                dns.zone: primary
            - broker: 2
              annotations:
                external-dns.alpha.kubernetes.io/hostname: kafka-2.example.com.
              labels:
                dns.zone: primary
```

This works, but it becomes repetitive and error-prone as clusters grow or when node IDs are not simple `0,1,2` sequences.
That is especially visible in KRaft clusters using node pools, where users often need to think in terms of node IDs and pod names rather than replica ordinals.

## Motivation

In most real deployments, per-broker annotations and labels follow a predictable pattern.
Examples include:

* `external-dns.alpha.kubernetes.io/hostname: kafka-{nodeId}.example.com.`
* `my-company.io/pod-name: {nodePodName}`
* `my-company.io/broker-id: {nodeId}`

> Requiring users to repeat these values broker by broker has several drawbacks:

Reviewer 2: ```suggestion
Requiring users to repeat these values for each broker has several drawbacks:
```

  example.com/node-id: "12"
  example.com/pod-name: my-cluster-brokers-12
```

### Where the templates apply

The new template fields apply only to per-broker resources.
They do not affect bootstrap resources.

They should be used everywhere Strimzi currently consumes per-broker annotations and labels from listener configuration, namely:

* per-broker `Service` resources created for `loadbalancer`, `nodeport`, and `cluster-ip` listeners
* per-broker `Route` resources
* per-broker `Ingress` resources
* per-broker `TLSRoute` resources, when that support is implemented

Bootstrap metadata remains configured explicitly through `configuration.bootstrap.annotations` and `configuration.bootstrap.labels`.

### Conflict handling and precedence

Users should be able to define a common default while also having a simple way to fully customize specific brokers.
To keep the behavior predictable and to allow users to omit values coming from the template, the precedence should be:

1. if `configuration.brokers[].annotations` is set for a broker, use it as the full annotation map for that broker
2. otherwise, if `configuration.perBrokerAnnotationsTemplate` is set, render it for that broker

> 3. if neither is set, use no broker annotations

Reviewer 2: ```suggestion
3. if neither is set, do not apply broker annotations
```

            external-dns.alpha.kubernetes.io/hostname: kafka-{nodeId}.example.com.
            external-dns.alpha.kubernetes.io/ttl: "60"
          perBrokerLabelsTemplate:
            exposure-type: standard
            example.com/node-id: "{nodeId}"
          brokers:
            - broker: 5
              annotations:
                external-dns.alpha.kubernetes.io/hostname: vip-broker.example.com.
              labels:
                exposure-type: premium
```

Because broker `5` defines explicit annotations and labels, it does not inherit any template-provided metadata.
This keeps the common case short while preserving precise control where needed.

## Affected/not affected projects

Affected:

* `strimzi-kafka-operator` API model
* `strimzi-kafka-operator` Cluster Operator listener validation and resource generation
* generated CRD documentation for listener configuration
* user-facing listener documentation
* future `TLSRoute` support for listeners

Not affected:

* Topic Operator behavior
* User Operator behavior
* Kafka protocol behavior itself
* bootstrap listener metadata behavior

## Compatibility

This proposal is fully backwards compatible.

> * Existing configurations using only `brokers[].annotations` and `brokers[].labels` continue to work unchanged.

Reviewer 2: ```suggestion
* Existing configurations using `brokers[].annotations` and `brokers[].labels` continue to work unchanged.
```
